src/hotspot/share/runtime/thread.cpp

```diff
@@ -915,8 +915,11 @@ void Thread::print_on(outputStream* st, bool print_extended_info) const {
 
     st->print("tid=" INTPTR_FORMAT " ", p2i(this));
     osthread()->print_on(st);
+
+    if (osthread()->get_state() != ZOMBIE) {
```

Comment:
I'm not sure print_on(), as opposed to print_on_error() can ever be called with a ZOMBIE thread. I didn't expect any change in this method.

Reply:
For thread, e.g. G1ConcurrentMarkThread, there is nothing to prevent calling _cm_thread->print_on(tty) after it terminated, although, I can not find a case right now.

You prefer an assertion instead?

Reply:
If Thread::print_on() (and ThreadsSMRSupport::print_info_on(Thread..)) cannot be called for a Zombie thread, I'd prefer an assertion testing that.

Reply:
I think this part is not necessary. Please see my general comment about
the thread closures.